### ISSUE TITLE
FITTER: Massive memory usage for crappy data

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#19]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/19) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Executing the following code on my imac on 3ca661ae698857742200307423eca08c56129639 causes the memory usage of the python process in activity tracker to grow very rapidly. I hit 20 GB within 30 seconds in one case. If I let it keep going eventually my computer becomes nearly unresponsive as I have not RAM left.
```python
import mass       
import numpy as np
data = np.array([1, 3, 1, 0, 1, 1, 3, 2, 2, 0, 1, 3, 0, 2, 8, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 2, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0])
pulseheights = np.array([ 13425.76458963,  13429.03365673,  13432.30272384,  13435.57179095,
        13438.84085806,  13442.10992516,  13445.37899227,  13448.64805938,
        13451.91712649,  13455.18619359,  13458.4552607 ,  13461.72432781,
        13464.99339492,  13468.26246202,  13471.53152913,  13474.80059624,
        13478.06966335,  13481.33873045,  13484.60779756,  13487.87686467,
        13491.14593178,  13494.41499888,  13497.68406599,  13500.9531331 ,
        13504.22220021,  13507.49126731,  13510.76033442,  13514.02940153,
        13517.29846863,  13520.56753574,  13523.83660285,  13527.10566996,
        13530.37473706,  13533.64380417,  13536.91287128,  13540.18193839,
        13543.45100549,  13546.7200726 ,  13549.98913971,  13553.25820682,
        13556.52727392,  13559.79634103,  13563.06540814,  13566.33447525,
        13569.60354235,  13572.87260946,  13576.14167657,  13579.41074368,
        13582.67981078,  13585.94887789,  13589.217945  ,  13592.48701211,
        13595.75607921,  13599.02514632,  13602.29421343,  13605.56328054,
        13608.83234764,  13612.10141475,  13615.37048186,  13618.63954896,
        13621.90861607,  13625.17768318,  13628.44675029,  13631.71581739,
        13634.9848845 ])
params_guess = [16.462571559496993, 13471.993702460593, 1.6462571559496992, None, None, None, None, None]
hold = [2]

fitter = mass.calibration.FeKBetaFitter()
fitter.fit(data, pulseheights, params_guess, hold=hold)
```

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Galen: I just noticed that I had a handful (3 or 4) local commits never pushed to the Bucket, so I merged them with master and pushed. (I understand that this week you've been using a new branch, but you should experiment with them.  Commit 5a96bb1 in particular is interesting, in that it limits the range of the exponential tail parameter to be no longer than 200 eV. That has got to put a limit on the time and memory consumed in fits.

On the other hand, I do reproduce your problem using the example code included above. I don't have bright ideas, but I'll see what I can do next week. Let me know what you come up with!

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Fluorescence line fitter: more appropriate bounds on variables.

Fixes #19, in which fitter can use an arbitrarily large number of "hidden bins" for the DFT-based Gaussian smearing algorithm.
